File: bindings/node/retrace.js
/*
 * retrace -- the Node runtime agent.
 *
 * A THIRD-PARTY implementation of the retrace supervisor
 * protocol (the same RTRD framing and state machine the
 * conformance suite's reference stub walks -- the third
 * runtime adapter after pyretrace and jretrace, three hook
 * systems, one protocol). diagnostics_channel gives the
 * runtime's own syscall-ish boundary -- file access, sockets,
 * child processes -- attributed as the runtime lane.
 *
 *   const retrace = require('retrace')
 *   retrace.supervise()        // joins RETRACE_SUPERVISOR env
 *
 * Zero-mandatory-config: with the env absent, supervise() is a
 * no-op (the preload plane's own gating doctrine). Events are
 * source=runtime; the labeler keeps libc / kernel / runtime as
 * lanes of one session.
 */

'use strict';

const fs = require('fs');
const net = require('net');
const os = require('os');

const MAGIC = Buffer.from('RTRD');
/* frozen v1 table (see src/supervisor/protocol.h; the
 * conformance suite asserts these never drift) */
const HELLO = 1, HEARTBEAT = 2, EVENT = 4, BYE = 6, WELCOME = 16;

const PIPE_PREFIX = '\\\\.\\pipe\\';

const state = {
  sock: null,        /* fd (POSIX) | handle (Windows) */
  agentId: 'pending',
  seq: 0,
  stop: false,
  heartbeat: null,
  isPipe: false,
};

function frame(mid, payload) {
  const body = Buffer.from(JSON.stringify(payload));
  const hdr = Buffer.alloc(12);
  MAGIC.copy(hdr, 0);
  hdr.writeUInt16LE(1, 4);
  hdr.writeUInt16LE(mid, 6);
  hdr.writeUInt32LE(body.length, 8);
  return Buffer.concat([hdr, body]);
}

function send(mid, payload) {
  if (!state.sock)
    return;
  const buf = frame(mid, payload);
  try {
    if (state.isPipe)
      fs.writeSync(state.sock, buf);
    else
      state.sock.write(buf);
  } catch (_) {
    state.sock = null;
  }
}

function emit(name, attrs) {
  state.seq += 1;
  send(EVENT, {
    agent_id: state.agentId,
    seq: state.seq,
    ts: Math.floor(Date.now() / 1000),
    name: name,
    attrs: Object.fromEntries(
      Object.entries(attrs || {}).map(([k, v]) => [k, String(v)])),
    source: 'runtime',
  });
}

/* ---- the runtime's own boundary (best-effort: the
 * diagnostics_channel surface churned across Node versions,
 * so each subscription is guarded and its absence is not an
 * error -- the direct emit() API is the stable floor) ---- */
function observeRuntime() {
  let dc = null;
  try {
    dc = require('diagnostics_channel');
  } catch (_) {
    return;
  }
  const on = (ch, name, pick) => {
    try {
      dc.subscribe(ch, (m) => {
        const attrs = pick(m);
        if (attrs)
          emit(name, attrs);
      });
    } catch (_) { /* channel absent on this runtime */ }
  };
  /* the fs channel surface: Node 18's 'fs.sync'/'fs.async'
   * carry {path}; Node 22+ split per-operation with the path
   * on the request -- every spelling is subscribed, guarded,
   * and read for either shape */
  const fsPick = (m) => {
    const p = m && (m.path ||
      (m.request && m.request.path) ||
      (m.response && m.response.path));
    return p ? { path: p } : null;
  };
  const fsRead = (m) => {
    const a = fsPick(m);
    return a;
  };
  for (const ch of ['fs.sync', 'fs.async', 'fs.sync.read',
    'fs.async.read'])
    on(ch, 'node.file.read', fsRead);
  for (const ch of ['fs.sync.write', 'fs.async.write'])
    on(ch, 'node.file.write', fsPick);
  on('net.client.socket', 'node.net.connect', () => ({}));
  on('child_process.run', 'node.exec.spawn', () => ({}));
}

function readWelcome(sockPath, nonce) {
  return new Promise((resolve, reject) => {
    const finish = (sock, isPipe) => {
      const chunks = [];
      const take = () => {
        const buf = Buffer.concat(chunks);
        if (buf.length < 12)
          return false;
        const len = buf.readUInt32LE(8);
        if (buf.length < 12 + len)
          return false;
        const mid = buf.readUInt16LE(6);
        const payload =
          JSON.parse(buf.subarray(12, 12 + len).toString());
        if (mid !== WELCOME) {
          reject(new Error('expected WELCOME, got ' + mid));
          return true;
        }
        resolve({ sock, isPipe, payload });
        return true;
      };
      if (isPipe) {
        let acc = Buffer.alloc(0);
        const step = () => {
          let got;
          let b;
          try {
            b = Buffer.alloc(4096);
            got = fs.readSync(sock, b, 0, b.length, null);
            if (got <= 0)
              throw new Error('pipe closed');
          } catch (e) {
            reject(e);
            return;
          }
          acc = Buffer.concat([acc, b.subarray(0, got)]);
          chunks.push(b.subarray(0, got));
          if (!take())
            setImmediate(step);
        };
        /* HELLO first */
        try {
          fs.writeSync(sock, frame(HELLO, {
            session_token: process.env.RETRACE_SESSION || '',
            nonce: nonce,
            pid: process.pid,
            ppid: 0,
            boot_id: 'noderuntime',
            cmdline: process.argv.slice(0, 4).join(' '),
            retrace_version: 'noderetrace-1',
          }));
        } catch (e) {
          reject(e);
          return;
        }
        step();
      } else {
        sock.on('data', (b) => {
          chunks.push(b);
          take();
        });
        sock.on('error', reject);
        sock.write(frame(HELLO, {
          session_token: process.env.RETRACE_SESSION || '',
          nonce: nonce,
          pid: process.pid,
          ppid: 0,
          boot_id: 'noderuntime',
          cmdline: process.argv.slice(0, 4).join(' '),
          retrace_version: 'noderetrace-1',
        }));
      }
    };
    if (sockPath.startsWith(PIPE_PREFIX)) {
      fs.open(sockPath, 'r+', (err, fd) => {
        if (err) {
          reject(err);
          return;
        }
        finish(fd, true);
      });
    } else {
      const s = net.connect({ path: sockPath });
      s.on('connect', () => finish(s, false));
      s.on('error', reject);
    }
  });
}

async function supervise() {
  if (process.env.RETRACE_SUPERVISOR !== '1')
    return false;
  const sockPath = process.env.RETRACE_SUPERVISOR_SOCK || '';
  const nonce = process.env.RETRACE_SUPERVISOR_NONCE || '';
  if (!sockPath)
    return false;

  let backoff = 500;
  while (!state.stop) {
    try {
      const r = await readWelcome(sockPath, nonce);
      state.sock = r.sock;
      state.isPipe = r.isPipe;
      state.agentId = r.payload.agent_id || 'pending';
      break;
    } catch (_) {
      await new Promise((res) => setTimeout(res, backoff));
      backoff = Math.min(backoff * 2, 10000);
    }
  }

  state.heartbeat = setInterval(() => {
    send(HEARTBEAT, { agent_id: state.agentId, seq: state.seq });
  }, 1000);

  observeRuntime();

  const bye = () => {
    if (state.stop)
      return;
    state.stop = true;
    clearInterval(state.heartbeat);
    send(BYE, { agent_id: state.agentId });
    try {
      if (state.isPipe)
        fs.closeSync(state.sock);
      else
        state.sock.end();
    } catch (_) { /* teardown best-effort */ }
  };
  process.on('exit', bye);
  process.on('SIGTERM', () => { bye(); process.exit(0); });
  process.on('SIGINT', () => { bye(); process.exit(0); });
  return true;
}

module.exports = { supervise, emit };

File: bindings/go/retrace/retrace.go
// Package retrace -- the Go runtime agent (TODO.impl/12).
//
// A THIRD-PARTY implementation of the retrace supervisor
// protocol (the same framing and state machine the conformance
// suite's reference stub and the Python/Node/JVM siblings
// walk). The cgo-free hook surface is the runtime's own
// boundaries: an http.RoundTripper wrapper names each request
// with the layer a libc interposer sees only as "a connect
// from pid N"; Emit carries anything the host program wants to
// attribute.
//
//	retrace.Supervise()          // joins RETRACE_SUPERVISOR env
//	retrace.HookHTTP()           // wraps http.DefaultTransport
//	retrace.Emit("go.marker", map[string]string{"k": "v"})
//
// Zero-mandatory-config: with the env absent, Supervise is a
// no-op (the preload plane's own gating doctrine). The agent
// registers as source=runtime; the labeler keeps libc / kernel
// / runtime as lanes of one session.
//
// Wire contract: RTRD framing, the frozen v1 message table
// (see src/supervisor/protocol.h; the conformance suite pins
// both sides to the same artifacts).
package retrace

import (
	"encoding/binary"
	"encoding/json"
	"io"
	"net"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"
)

const (
	magic   = "RTRD"
	verV1   = 1
	mHello  = 1
	mBeat   = 2
	mEvent  = 4
	mBye    = 6
	mWelc   = 16
	beatDur = time.Second
)

type state struct {
	mu      sync.Mutex
	sock    string
	nonce   string
	conn    io.ReadWriteCloser
	agentID string
	seq     uint64
	stop    chan struct{}
	done    chan struct{}
	hooked  bool
}

var st state

// Supervise joins the supervisor named by RETRACE_SUPERVISOR;
// a no-op when absent. Returns whether the agent armed.
func Supervise() bool {
	if os.Getenv("RETRACE_SUPERVISOR") != "1" {
		return false
	}
	sock := os.Getenv("RETRACE_SUPERVISOR_SOCK")
	if sock == "" {
		return false
	}
	st.mu.Lock()
	if st.stop != nil {
		st.mu.Unlock()
		return true
	}
	st.sock = sock
	st.nonce = os.Getenv("RETRACE_SUPERVISOR_NONCE")
	st.stop = make(chan struct{})
	done := make(chan struct{})
	st.done = done
	st.mu.Unlock()
	go agentLoop()
	// settle past HELLO/WELCOME so early emits carry the id
	deadline := time.Now().Add(3 * time.Second)
	for time.Now().Before(deadline) {
		st.mu.Lock()
		ready := st.conn != nil
		st.mu.Unlock()
		if ready {
			break
		}
		select {
		case <-done:
			return false
		case <-time.After(50 * time.Millisecond):
		}
	}
	return true
}

// Shutdown sends BYE and closes the connection (the graceful
// path; without it the daemon's liveness sweep reaps the seat).
func Shutdown() {
	st.mu.Lock()
	stop := st.stop
	st.mu.Unlock()
	if stop == nil {
		return
	}
	select {
	case <-stop:
	default:
		close(stop)
	}
	st.mu.Lock()
	done := st.done
	st.mu.Unlock()
	if done != nil {
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

// pipeConn adapts a byte-mode named pipe opened as a file to
// the socket surface the agent speaks (Windows Go has no
// AF_UNIX; the pipe carries the same RTRD framing).
type pipeConn struct {
	f *os.File
}

func (p pipeConn) Read(b []byte) (int, error)  { return p.f.Read(b) }
func (p pipeConn) Write(b []byte) (int, error) { return p.f.Write(b) }
func (p pipeConn) Close() error                { return p.f.Close() }

func dial(sock string) (io.ReadWriteCloser, error) {
	if strings.HasPrefix(sock, `\\.\pipe\`) {
		return pipeConn{}, nil // placeholder; replaced below
	}
	return net.Dial("unix", sock)
}

func dialPipe(sock string) (io.ReadWriteCloser, error) {
	f, err := os.OpenFile(sock, os.O_RDWR, 0)
	if err != nil {
		return nil, err
	}
	return pipeConn{f: f}, nil
}

func frame(mid uint16, payload []byte) []byte {
	hdr := make([]byte, 12)
	copy(hdr, magic)
	binary.LittleEndian.PutUint16(hdr[4:], verV1)
	binary.LittleEndian.PutUint16(hdr[6:], mid)
	binary.LittleEndian.PutUint32(hdr[8:], uint32(len(payload)))
	return append(hdr, payload...)
}

func send(mid uint16, obj any) {
	st.mu.Lock()
	c := st.conn
	st.mu.Unlock()
	if c == nil {
		return
	}
	b, err := json.Marshal(obj)
	if err != nil {
		return
	}
	if _, err := c.Write(frame(mid, b)); err != nil {
		st.mu.Lock()
		if st.conn == c {
			c.Close()
			st.conn = nil
		}
		st.mu.Unlock()
	}
}

func recvExact(c io.Reader, n int) ([]byte, error) {
	buf := make([]byte, n)
	if _, err := io.ReadFull(c, buf); err != nil {
		return nil, err
	}
	return buf, nil
}

func hello(sock, nonce string) (io.ReadWriteCloser, string, error) {
	var c io.ReadWriteCloser
	var err error
	if strings.HasPrefix(sock, `\\.\pipe\`) {
		c, err = dialPipe(sock)
	} else {
		c, err = dial(sock)
	}
	if err != nil {
		return nil, "", err
	}
	hb, _ := json.Marshal(map[string]any{
		"session_token": os.Getenv("RETRACE_SESSION"),
		"nonce":         nonce,
		"pid":           os.Getpid(),
		"ppid":          os.Getppid(),
		"boot_id":       "goruntime",
		"cmdline":       strings.Join(os.Args[:min(4, len(os.Args))], " "),
		"retrace_version": "goretrace-1",
	})
	if _, err := c.Write(frame(mHello, hb)); err != nil {
		c.Close()
		return nil, "", err
	}
	// WELCOME (a POLICY_SET may trail; runtime agents are
	// observers -- anything after the welcome is drained by
	// the beat loop's ignore-reader)
	if dl, ok := c.(interface{ SetReadDeadline(time.Time) error }); ok {
		dl.SetReadDeadline(time.Now().Add(5 * time.Second))
		defer dl.SetReadDeadline(time.Time{})
	}
	hdr, err := recvExact(c, 12)
	if err != nil {
		c.Close()
		return nil, "", err
	}
	if string(hdr[:4]) != magic {
		c.Close()
		return nil, "", io.ErrUnexpectedEOF
	}
	ln := binary.LittleEndian.Uint32(hdr[8:])
	body, err := recvExact(c, int(ln))
	if err != nil {
		c.Close()
		return nil, "", err
	}
	if binary.LittleEndian.Uint16(hdr[6:]) != mWelc {
		c.Close()
		return nil, "", io.ErrUnexpectedEOF
	}
	var w struct {
		AgentID string `json:"agent_id"`
	}
	json.Unmarshal(body, &w)
	return c, w.AgentID, nil
}

func agentLoop() {
	defer close(st.done)
	backoff := 500 * time.Millisecond
	for {
		st.mu.Lock()
		stopped := st.stop
		st.mu.Unlock()
		if stopped == nil {
			return
		}
		select {
		case <-stopped:
			return
		default:
		}
		c, id, err := hello(st.sock, st.nonce)
		if err != nil {
			select {
			case <-stopped:
				return
			case <-time.After(backoff):
			}
			if backoff < 10*time.Second {
				backoff *= 2
			}
			continue
		}
		backoff = 500 * time.Millisecond
		st.mu.Lock()
		st.conn = c
		st.agentID = id
		st.mu.Unlock()
		go drain(c) // daemon pushes (policy/ping) are read+dropped
		for {
			st.mu.Lock()
			dead := st.conn == nil
			st.mu.Unlock()
			if dead {
				break // drain saw EOF; reconnect
			}
			select {
			case <-stopped:
				send(mBye, map[string]any{"agent_id": st.agentID})
				st.mu.Lock()
				if st.conn != nil {
					st.conn.Close()
					st.conn = nil
				}
				st.mu.Unlock()
				return
			case <-time.After(beatDur):
				st.mu.Lock()
				seq := st.seq
				st.mu.Unlock()
				send(mBeat, map[string]any{
					"agent_id": st.agentID,
					"seq":      seq,
				})
			}
		}
	}
}

func drain(c io.ReadWriteCloser) {
	buf := make([]byte, 4096)
	for {
		if _, err := c.Read(buf); err != nil {
			st.mu.Lock()
			if st.conn == c {
				st.conn = nil
			}
			st.mu.Unlock()
			return
		}
	}
}

// Emit sends one runtime-attributed event (public API).
func Emit(name string, attrs map[string]string) {
	st.mu.Lock()
	st.seq++
	seq := st.seq
	agent := st.agentID
	st.mu.Unlock()
	payload := map[string]any{
		"agent_id": agent,
		"seq":      seq,
		"ts":       time.Now().Unix(),
		"name":     name,
		"attrs":    attrs,
		"source":   "runtime",
	}
	send(mEvent, payload)
}

// hookRT wraps an http.RoundTripper with per-request events.
type hookRT struct {
	next http.RoundTripper
}

func (h hookRT) RoundTrip(req *http.Request) (*http.Response, error) {
	Emit("go.http.request", map[string]string{
		"method": req.Method,
		"host":   req.URL.Host,
		"path":   req.URL.Path,
	})
	resp, err := h.next.RoundTrip(req)
	if err != nil {
		Emit("go.http.error", map[string]string{
			"host": req.URL.Host, "error": err.Error()})
		return resp, err
	}
	Emit("go.http.response", map[string]string{
		"host":   req.URL.Host,
		"path":   req.URL.Path,
		"status": itoa(resp.StatusCode),
	})
	return resp, nil
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b [8]byte
	i := len(b)
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}

// HookHTTP wraps http.DefaultTransport with the observing
// RoundTripper (idempotent).
func HookHTTP() {
	st.mu.Lock()
	defer st.mu.Unlock()
	if st.hooked {
		return
	}
	st.hooked = true
	t, ok := http.DefaultTransport.(http.RoundTripper)
	if !ok || t == nil {
		t = http.DefaultTransport
	}
	http.DefaultTransport = hookRT{next: t}
}
